src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicSession.cs

```diff
@@ -0,0 +1,149 @@
+                                    options.MaxBidirectionalStreams,
+                                    options.MaxUnidirectionalStreams);
+            }
+
+            MsQuicStatusException.ThrowIfFailed(MsQuicApi.Api._listenerOpenDelegate(
+                _nativeObjPtr,
+                MsQuicListener.NativeCallbackHandler,
+                IntPtr.Zero,
+                out IntPtr listenerPointer));
+
+            return listenerPointer;
+        }
+
+        // TODO call this for graceful shutdown?
+        public void ShutDown(
+            QUIC_CONNECTION_SHUTDOWN_FLAG Flags,
+            ushort ErrorCode)
+        {
+            MsQuicApi.Api._sessionShutdownDelegate(
+                _nativeObjPtr,
+                (uint)Flags,
+                ErrorCode);
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        public void SetPeerBiDirectionalStreamCount(ushort count)
+        {
+            SetUshortParamter(QUIC_PARAM_SESSION.PEER_BIDI_STREAM_COUNT, count);
+        }
+
+        public void SetPeerUnidirectionalStreamCount(ushort count)
+        {
+            SetUshortParamter(QUIC_PARAM_SESSION.PEER_UNIDI_STREAM_COUNT, count);
+        }
+
+        private unsafe void SetUshortParamter(QUIC_PARAM_SESSION param, ushort count)
+        {
+            var buffer = new MsQuicNativeMethods.QuicBuffer()
+            {
+                Length = sizeof(ushort),
+                Buffer = (byte*)&count
+            };
+
+            SetParam(param, buffer);
+        }
+
+        public void SetDisconnectTimeout(TimeSpan timeout)
+        {
+            SetULongParamter(QUIC_PARAM_SESSION.DISCONNECT_TIMEOUT, (ulong)timeout.TotalMilliseconds);
+        }
+
+        public void SetIdleTimeout(TimeSpan timeout)
+        {
+            SetULongParamter(QUIC_PARAM_SESSION.IDLE_TIMEOUT, (ulong)timeout.TotalMilliseconds);
+
+        }
+        private unsafe void SetULongParamter(QUIC_PARAM_SESSION param, ulong count)
+        {
+            var buffer = new MsQuicNativeMethods.QuicBuffer()
+            {
+                Length = sizeof(ulong),
+                Buffer = (byte*)&count
+            };
+            SetParam(param, buffer);
+        }
+
+        private void SetParam(
+          QUIC_PARAM_SESSION param,
+          MsQuicNativeMethods.QuicBuffer buf)
+        {
+            MsQuicStatusException.ThrowIfFailed(MsQuicApi.Api.UnsafeSetParam(
+                _nativeObjPtr,
+                (uint)QUIC_PARAM_LEVEL.SESSION,
+                (uint)param,
+                buf));
+        }
+
+        ~MsQuicSession()
+        {
+            Dispose(false);
+        }
+
+        private void Dispose(bool disposing)
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            MsQuicApi.Api._sessionCloseDelegate?.Invoke(_nativeObjPtr);
```

Comment:
using a managed reference in finalizer.

Reply:
So the MsQuicApi is effectively a static singleton (with lazy initialization). So I think it should always be available. Are there other concerns about using a managed reference in a finalizer?